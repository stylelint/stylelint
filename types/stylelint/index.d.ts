declare module 'stylelint' {
	import type * as PostCSS from 'postcss';
	import type { GlobbyOptions } from 'globby';
	import type { cosmiconfig } from 'cosmiconfig';

	namespace stylelint {
		export type Severity = 'warning' | 'error';

		export type ConfigExtends = string | string[];
		export type ConfigPlugins = string | string[];
		export type ConfigProcessor = string | [string, Object];
		export type ConfigProcessors = string | ConfigProcessor[];
		export type ConfigIgnoreFiles = string | string[];
		export type ConfigRuleSettings<T, O extends Object> =
			| null
			| undefined
			| NonNullable<T>
			| [NonNullable<T>]
			| [NonNullable<T>, O];
		export type ConfigRules = {
			[ruleName: string]: ConfigRuleSettings<any, Object>;
		};
		export type ConfigOverride = Omit<Config, 'overrides'> & {
			files: string | string[];
		};

		export type DisableOptions = {
			except?: (string | RegExp)[];
			severity?: Severity;
		};
		export type DisableSettings = ConfigRuleSettings<boolean, DisableOptions>;

		export type ResultProcessor = (result: LintResult, file: string | undefined) => LintResult;

		export type Config = {
			extends?: ConfigExtends;
			plugins?: ConfigPlugins;
			pluginFunctions?: {
				[pluginName: string]: Rule;
			};
			processors?: ConfigProcessors;
			processorFunctions?: Function[];
			ignoreFiles?: ConfigIgnoreFiles;
			ignorePatterns?: string;
			rules?: ConfigRules;
			codeProcessors?: CodeProcessor[];
			resultProcessors?: ResultProcessor[];
			quiet?: boolean;
			defaultSeverity?: Severity;
			ignoreDisables?: DisableSettings;
			reportNeedlessDisables?: DisableSettings;
			reportInvalidScopeDisables?: DisableSettings;
			reportDescriptionlessDisables?: DisableSettings;
			overrides?: ConfigOverride[];
			customSyntax?: CustomSyntax;
		};

		// A meta-type that returns a union over all properties of `T` whose values
		// have type `U`.
		type PropertyNamesOfType<T, U> = {
			[K in keyof T]-?: T[K] extends U ? K : never;
		}[keyof T];
		export type DisablePropertyName = PropertyNamesOfType<Config, DisableSettings>;

		// This type has the same properties as `CosmiconfigResult` from `cosmiconfig`.
		export type CosmiconfigResult = {
			config: Config;
			filepath: string;
			isEmpty?: boolean;
		} | null;

		export type ConfigurationError = Error & { code: 78 };

		export type DisabledRange = {
			comment: PostCSS.Comment;
			start: number;
			strictStart: boolean;
			end?: number;
			strictEnd?: boolean;
			rules?: string[];
			description?: string;
		};

		export type DisabledRangeObject = {
			[ruleName: string]: DisabledRange[];
		};

		export type DisabledWarning = { line: number; rule: string };

		type FileCache = {
			calcHashOfConfig: (config: Config) => void;
			hasFileChanged: (absoluteFilepath: string) => boolean;
			reconcile: () => void;
			destroy: () => void;
			removeEntry: (absoluteFilepath: string) => void;
		};

		export type StylelintPostcssResult = {
			ruleSeverities: { [ruleName: string]: Severity };
			customMessages: { [ruleName: string]: RuleMessage };
			ruleMetadata: { [ruleName: string]: Partial<RuleMeta> };
			quiet?: boolean;
			disabledRanges: DisabledRangeObject;
			disabledWarnings?: DisabledWarning[];
			ignored?: boolean;
			stylelintError?: boolean;
			stylelintWarning?: boolean;
			disableWritingFix?: boolean;
			config?: Config;
			ruleDisableFix?: boolean;
		};

		type EmptyResult = {
			root: {
				nodes?: undefined;
				source: {
					lang?: undefined;
					input: {
						file?: string;
					};
				};
			};
			messages: PostCSS.Message[];
			opts: undefined;
		};

		export type WarningOptions = PostCSS.WarningOptions & {
			stylelintType?: string;
			severity?: Severity;
			rule?: string;
		};

		export type PostcssResult = (PostCSS.Result | EmptyResult) & {
			stylelint: StylelintPostcssResult;
			warn(message: string, options?: WarningOptions): void;
		};

		export type Formatter = (results: LintResult[], returnValue: LinterResult) => string;

		export type FormatterType =
			| 'compact'
			| 'github'
			| 'json'
			| 'string'
			| 'tap'
			| 'unix'
			| 'verbose';

		export type CustomSyntax = string | PostCSS.Syntax;

		// Note: With strict function types enabled, function signatures are checked contravariantly.
		// This means that it would not be possible for rule authors to narrow the message function
		// parameters to e.g. just `string`. Declaring the type for rule message functions through
		// method declarations tricks TypeScript into bivariant signature checking. More details can
		// be found here: https://stackoverflow.com/questions/52667959/what-is-the-purpose-of-bivariancehack-in-typescript-types.
		// and in the original discussion: https://github.com/stylelint/stylelint/pull/6147#issuecomment-1155337016.
		export type RuleMessageFunc = {
			bivariance(...args: (string | number | boolean | RegExp)[]): string;
		}['bivariance'];

		export type RuleMessage = string | RuleMessageFunc;

		export type RuleMessages = { [message: string]: RuleMessage };

		export type RuleOptionsPossibleFunc = (value: unknown) => boolean;

		export type RuleOptionsPossible = boolean | number | string | RuleOptionsPossibleFunc;

		export type RuleOptions = {
			actual: unknown;
			possible?:
				| RuleOptionsPossibleFunc
				| RuleOptionsPossible[]
				| Record<string, RuleOptionsPossible[]>;
			optional?: boolean;
		};

		export type RuleContext = {
			fix?: boolean | undefined;
			newline?: string | undefined;
		};

		/**
		 * @deprecated Use `RuleContext` instead.
		 */
		export type PluginContext = RuleContext;

		export type RuleBase<P = any, S = any> = (
			primaryOption: P,
			secondaryOptions: Record<string, S>,
			context: RuleContext,
		) => (root: PostCSS.Root, result: PostcssResult) => Promise<void> | void;

		export type RuleMeta = {
			url: string;
			deprecated?: boolean;
			fixable?: boolean;
		};

		export type Rule<P = any, S = any> = RuleBase<P, S> & {
			ruleName: string;
			messages: RuleMessages;
			primaryOptionArray?: boolean;
			meta?: RuleMeta;
		};

		export type Plugin = RuleBase;

		export type CodeProcessor = (code: string, file: string | undefined) => string;

		export type GetPostcssOptions = {
			code?: string;
			codeFilename?: string;
			filePath?: string;
			codeProcessors?: CodeProcessor[];
			syntax?: string;
			customSyntax?: CustomSyntax;
		};

		export type GetLintSourceOptions = GetPostcssOptions & {
			existingPostcssResult?: PostCSS.Result;
			cache?: boolean;
		};

		export type LinterOptions = {
			files?: string | string[];
			globbyOptions?: GlobbyOptions;
			cache?: boolean;
			cacheLocation?: string;
			cacheStrategy?: string;
			code?: string;
			codeFilename?: string;
			config?: Config;
			configFile?: string;
			configBasedir?: string;
			/**
			 * The working directory to resolve files from. Defaults to the
			 * current working directory.
			 */
			cwd?: string;
			ignoreDisables?: boolean;
			ignorePath?: string | string[];
			ignorePattern?: string[];
			reportDescriptionlessDisables?: boolean;
			reportNeedlessDisables?: boolean;
			reportInvalidScopeDisables?: boolean;
			maxWarnings?: number;
			/** @deprecated Use `customSyntax` instead. Using this option will result in an error. */
			syntax?: string;
			customSyntax?: CustomSyntax;
			formatter?: FormatterType | Formatter;
			disableDefaultIgnores?: boolean;
			fix?: boolean;
			allowEmptyInput?: boolean;
			quiet?: boolean;
		};

		export type CssSyntaxError = {
			file?: string;
			input: {
				column: number;
				file?: string;
				line: number;
				source: string;
			};
			/**
			 * The line of the inclusive start position of the error.
			 */
			line: number;
			/**
			 * The column of the inclusive start position of the error.
			 */
			column: number;
			/**
			 * The line of the exclusive end position of the error.
			 */
			endLine?: number;
			/**
			 * The column of the exclusive end position of the error.
			 */
			endColumn?: number;
			message: string;
			name: string;
			reason: string;
			source: string;
		};

		export type Warning = {
			/**
			 * The line of the inclusive start position of the warning.
			 */
			line: number;
			/**
			 * The column of the inclusive start position of the warning.
			 */
			column: number;
			/**
			 * The line of the exclusive end position of the warning.
			 */
			endLine?: number;
			/**
			 * The column of the exclusive end position of the warning.
			 */
			endColumn?: number;
			rule: string;
			severity: Severity;
			text: string;
			stylelintType?: string;
		};

		export type LintResult = {
			source?: string;
			deprecations: {
				text: string;
				reference: string;
			}[];
			invalidOptionWarnings: {
				text: string;
			}[];
			parseErrors: (PostCSS.Warning & { stylelintType: string })[];
			errored?: boolean;
			warnings: Warning[];
			ignored?: boolean;
			/**
			 * Internal use only. Do not use or rely on this property. It may
			 * change at any time.
			 * @internal
			 */
			_postcssResult?: PostcssResult;
		};

		export type DisableReportRange = {
			rule: string;
			start: number;
			end?: number;
		};

		export type RangeType = DisabledRange & { used?: boolean };

		export type DisableReportEntry = {
			source?: string;
			ranges: DisableReportRange[];
		};

		export type LinterResult = {
			/**
			 * The working directory from which the linter was run when the
			 * results were generated.
			 */
			cwd: string;
			results: LintResult[];
			errored: boolean;
			output: any;
			maxWarningsExceeded?: {
				maxWarnings: number;
				foundWarnings: number;
			};
			reportedDisables: DisableOptionsReport;
			descriptionlessDisables?: DisableOptionsReport;
			needlessDisables?: DisableOptionsReport;
			invalidScopeDisables?: DisableOptionsReport;
			/**
			 * Each rule metadata by name.
			 */
			ruleMetadata: { [ruleName: string]: Partial<RuleMeta> };
		};

		export type Problem = {
			ruleName: string;
			result: PostcssResult;
			message: RuleMessage;
			messageArgs?: Parameters<RuleMessageFunc> | undefined;
			node: PostCSS.Node;
			/**
			 * The inclusive start index of the problem, relative to the node's
			 * source text.
			 */
			index?: number;
			/**
			 * The exclusive end index of the problem, relative to the node's
			 * source text.
			 */
			endIndex?: number;
			/**
			 * The inclusive start position of the problem, relative to the
			 * node's source text. If provided, this will be used instead of
			 * `index`.
			 */
			start?: {
				line: number;
				column: number;
			};
			/**
			 * The exclusive end position of the problem, relative to the
			 * node's source text. If provided, this will be used instead of
			 * `endIndex`.
			 */
			end?: {
				line: number;
				column: number;
			};
			word?: string;
			line?: number;
			/**
			 * Optional severity override for the problem.
			 */
			severity?: Severity;
		};

		export type LonghandSubPropertiesOfShorthandProperties = ReadonlyMap<
			| 'animation'
			| 'background'
			| 'border'
			| 'border-block-end'
			| 'border-block-start'
			| 'border-bottom'
			| 'border-color'
			| 'border-image'
			| 'border-inline-end'
			| 'border-inline-start'
			| 'border-left'
			| 'border-radius'
			| 'border-right'
			| 'border-style'
			| 'border-top'
			| 'border-width'
			| 'column-rule'
			| 'columns'
			| 'flex'
			| 'flex-flow'
			| 'font'
			| 'grid'
			| 'grid-area'
			| 'grid-column'
			| 'grid-gap'
			| 'grid-row'
			| 'grid-template'
			| 'list-style'
			| 'margin'
			| 'mask'
			| 'outline'
			| 'padding'
			| 'text-decoration'
			| 'text-emphasis'
			| 'transition',
			ReadonlySet<string>
		>;

		export type PublicApi = PostCSS.PluginCreator<PostcssPluginOptions> & {
			/**
			 * Runs stylelint with the given options and returns a Promise that
			 * resolves to the results.
			 */
			lint: (options: LinterOptions) => Promise<LinterResult>;
			/**
			 * Available rules.
			 */
			rules: { [k: string]: Rule };
			/**
			 * Result report formatters by name.
			 */
			formatters: { [k: string]: Formatter };
			/**
			 * Creates a Stylelint plugin.
			 */
			createPlugin: (ruleName: string, rule: Rule) => { ruleName: string; rule: Rule };
			/**
			 * Internal use only. Do not use or rely on this method. It may
			 * change at any time.
			 * @internal
			 */
			createLinter: (options: LinterOptions) => InternalApi;
			/**
			 * Resolves the effective configuration for a given file. Resolves to
			 * `undefined` if no config is found.
			 * @param filePath - The path to the file to get the config for.
			 * @param options - The options to use when creating the Stylelint instance.
			 */
			resolveConfig: (
				filePath: string,
				options?: Pick<LinterOptions, 'cwd' | 'config' | 'configBasedir' | 'configFile'>,
			) => Promise<stylelint.Config | undefined>;
			utils: {
				/**
				 * Report a problem.
				 *
				 * This function accounts for `disabledRanges` attached to the result.
				 * That is, if the reported problem is within a disabledRange,
				 * it is ignored. Otherwise, it is attached to the result as a
				 * postcss warning.
				 *
				 * It also accounts for the rule's severity.
				 *
				 * You *must* pass *either* a node or a line number.
				 */
				report: (problem: Problem) => void;
				/**
				 * Given an object of problem messages, return another
				 * that provides the same messages postfixed with the rule
				 * that has been violated.
				 */
				ruleMessages: <T extends RuleMessages, R extends { [K in keyof T]: T[K] }>(
					ruleName: string,
					messages: T,
				) => R;
				/**
				 * Validate a rule's options.
				 *
				 * See existing rules for examples.
				 */
				validateOptions: (
					result: PostcssResult,
					ruleName: string,
					...optionDescriptions: RuleOptions[]
				) => boolean;
				/**
				 * Useful for third-party code (e.g. plugins) to run a PostCSS Root
				 * against a specific rule and do something with the warnings
				 */
				checkAgainstRule: <T, O extends Object>(
					options: {
						ruleName: string;
						ruleSettings: ConfigRuleSettings<T, O>;
						root: PostCSS.Root;
						result?: PostcssResult;
						context?: RuleContext;
					},
					callback: (warning: PostCSS.Warning) => void,
				) => void;
			};
			reference: {
				longhandSubPropertiesOfShorthandProperties: LonghandSubPropertiesOfShorthandProperties;
			};
		};

		/**
		 * Internal use only. Do not use or rely on this type. It may change at
		 * any time.
		 * @internal
		 */
		export type InternalApi = {
			_options: LinterOptions & { cwd: string };
			_extendExplorer: ReturnType<typeof cosmiconfig>;
			_specifiedConfigCache: Map<Config, Promise<CosmiconfigResult>>;
			_postcssResultCache: Map<string, PostCSS.Result>;
			_fileCache: FileCache;
		};

		export type DisableOptionsReport = DisableReportEntry[];

		export type PostcssPluginOptions = Omit<LinterOptions, 'syntax' | 'customSyntax'> | Config;
	}

	const stylelint: stylelint.PublicApi;

	export = stylelint;
}
