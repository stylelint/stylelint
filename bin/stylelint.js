#!/usr/bin/env node

'use strict';

// to use V8's code cache to speed up instantiation time
require('v8-compile-cache');

require('../lib/cli')(process.argv.slice(2));
