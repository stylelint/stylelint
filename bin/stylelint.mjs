#!/usr/bin/env node

import cli from '../lib/cli.mjs';

cli(process.argv.slice(2));
