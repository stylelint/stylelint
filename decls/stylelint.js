export type stylelint$configExtends = string | Array<string>;
export type stylelint$configPlugins = string | Array<string>;
export type stylelint$configProcessors =
  | string
  | Array<string | [string, Object]>;
export type stylelint$configIgnoreFiles = string | Array<string>;

export type stylelint$configRuleSettings = any | [any, Object];
export type stylelint$configRules = {
  [ruleName: string]: stylelint$configRuleSettings
};

export type stylelint$config = {
  extends?: stylelint$configExtends,
  plugins?: stylelint$configPlugins,
  pluginFunctions?: {
    [pluginName: string]: Function
  },
  processors?: stylelint$configProcessors,
  processorFunctions?: Array<Function>,
  ignoreFiles?: stylelint$configIgnoreFiles,
  ignorePatterns?: string,
  rules?: stylelint$configRules,
  codeProcessors?: Array<Function>,
  resultProcessors?: Array<Function>,
  quiet?: boolean
};

export type stylelint$syntaxes = "scss" | "less" | "sugarss";

export type stylelint$options = {
  config?: stylelint$config,
  configFile?: string,
  configBasedir?: string,
  configOverrides?: Object,
  ignoreDisables?: boolean,
  ignorePath?: string,
  reportNeedlessDisables?: boolean,
  syntax?: stylelint$syntaxes,
  customSyntax?: string,
  fix?: boolean
};

export type stylelint$internalApi = {
  _options: stylelint$options,
  _extendExplorer: { load: Function },
  _fullExplorer: { load: Function },
  _configCache: Map<string, Object>,
  _specifiedConfigCache: Map<string, Object>,
  _postcssResultCache: Map<string, Object>,

  _augmentConfig: Function,
  _getPostcssResult: Function,
  _lintSource: Function,
  _createStylelintResult: Function,
  _createEmptyPostcssResult?: Function,

  getConfigForFile: Function,
  isPathIgnored: Function,
  lintSource: Function
};

export type stylelint$warning = {
  line: number,
  column: number,
  rule: string,
  severity: string,
  text: string
};

export type stylelint$result = {
  source: string,
  deprecations: Array<{
    text: string,
    reference: string
  }>,
  invalidOptionWarnings: Array<{
    text: string
  }>,
  parseErrors: Array<stylelint$warning>,
  errored?: boolean,
  warnings: Array<stylelint$warning>,
  ignored?: boolean,
  _postcssResult?: Object
};

export type stylelint$needlessDisablesReport = Array<{
  source: string,
  ranges: Array<{
    source: string,
    ranges: Array<{
      start: number,
      end?: number
    }>
  }>
}>;

export type stylelint$standaloneReturnValue = {
  results: Array<stylelint$result>,
  errored: boolean,
  output: any,
  needlessDisables?: stylelint$needlessDisablesReport
};

export type stylelint$standaloneOptions = {
  files?: string | Array<string>,
  cache?: boolean,
  cacheLocation?: string,
  code?: string,
  codeFilename?: string,
  config?: stylelint$config,
  configFile?: string,
  configBasedir?: string,
  configOverrides?: Object,
  ignoreDisables?: boolean,
  ignorePath?: string,
  reportNeedlessDisables?: boolean,
  syntax?: stylelint$syntaxes,
  customSyntax?: string,
  formatter?: "json" | "string" | "verbose" | Function,
  disableDefaultIgnores?: boolean,
  allowEmptyInput?: boolean,
  fix?: boolean
};
