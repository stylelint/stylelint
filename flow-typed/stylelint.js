export type stylelint$configExtends = string | Array<string>;
export type stylelint$configPlugins = string | Array<string>;
export type stylelint$configProcessors =
  | string
  | Array<string | [string, Object]>;
export type stylelint$configIgnoreFiles = string | Array<string>;

export type stylelint$configRuleSettings = any | [any, Object];
export type stylelint$configRules = {
  [ruleName: string]: stylelint$configRuleSettings
};

export type stylelint$config = {
  extends?: stylelint$configExtends,
  plugins?: stylelint$configPlugins,
  pluginFunctions?: {
    [pluginName: string]: Function
  },
  processors?: stylelint$configProcessors,
  processorFunctions?: Array<Function>,
  ignoreFiles?: stylelint$configIgnoreFiles,
  ignorePatterns?: string,
  rules?: stylelint$configRules,
  codeProcessors?: Array<Function>,
  resultProcessors?: Array<Function>,
  quiet?: boolean,
  defaultSeverity?: string
};

export type stylelint$syntaxes = "scss" | "less" | "sugarss";

export type stylelint$options = {
  config?: stylelint$config,
  configFile?: string,
  configBasedir?: string,
  configOverrides?: Object,
  ignoreDisables?: boolean,
  ignorePath?: string,
  reportInvalidScopeDisables?: boolean,
  reportNeedlessDisables?: boolean,
  syntax?: stylelint$syntaxes,
  customSyntax?: string,
  fix?: boolean
};

export type stylelint$internalApi = {
  _options: stylelint$options,
  _extendExplorer: {
    search: string => Promise<null | Object>,
    load: string => Promise<null | Object>
  },
  _fullExplorer: {
    search: string => Promise<null | Object>,
    load: string => Promise<null | Object>
  },
  _configCache: Map<string, Object>,
  _specifiedConfigCache: Map<string, Object>,
  _postcssResultCache: Map<string, Object>,

  _augmentConfig: Function,
  _getPostcssResult: Function,
  _lintSource: Function,
  _createStylelintResult: Function,
  _createEmptyPostcssResult?: Function,

  getConfigForFile: Function,
  isPathIgnored: Function,
  lintSource: Function
};

export type stylelint$warning = {
  line: number,
  column: number,
  rule: string,
  severity: string,
  text: string
};

export type stylelint$result = {
  source: string,
  deprecations: Array<{
    text: string,
    reference: string
  }>,
  invalidOptionWarnings: Array<{
    text: string
  }>,
  parseErrors: Array<stylelint$warning>,
  errored?: boolean,
  warnings: Array<stylelint$warning>,
  ignored?: boolean,
  _postcssResult?: Object
};

export type stylelint$cssSyntaxError = {
  column: number,
  file?: string,
  input: {
    column: number,
    file?: string,
    line: number,
    source: string
  },
  line: number,
  message: string,
  name: string,
  reason: string,
  source: string
};

export type stylelint$disableOptionsReport = Array<{
  source: string,
  ranges: Array<{
    unusedRule: string,
    start: number,
    end?: number
  }>
}>;

export type stylelint$standaloneReturnValue = {
  results: Array<stylelint$result>,
  errored: boolean,
  output: any,
  maxWarningsExceeded?: {
    maxWarnings: number,
    foundWarnings: number
  },
  needlessDisables?: stylelint$disableOptionsReport,
  invalidScopeDisables?: stylelint$disableOptionsReport
};

export type stylelint$standaloneOptions = {
  files?: string | Array<string>,
  globbyOptions?: Object,
  cache?: boolean,
  cacheLocation?: string,
  code?: string,
  codeFilename?: string,
  config?: stylelint$config,
  configFile?: string,
  configBasedir?: string,
  configOverrides?: Object,
  printConfig?: string,
  ignoreDisables?: boolean,
  ignorePath?: string,
  ignorePattern?: RegExp,
  reportNeedlessDisables?: boolean,
  reportInvalidScopeDisables?: boolean,
  maxWarnings?: number,
  syntax?: stylelint$syntaxes,
  customSyntax?: string,
  formatter?: "compact" | "json" | "string" | "unix" | "verbose" | Function,
  disableDefaultIgnores?: boolean,
  fix?: boolean,
  allowEmptyInput: boolean
};
