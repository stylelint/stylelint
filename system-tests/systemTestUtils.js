"use strict";

const _ = require("lodash");
const path = require("path");
const stripAnsi = require("strip-ansi");

function caseFilePath(caseNumber, fileName) {
  return path.join(__dirname, caseNumber, fileName);
}

function caseStylesheetGlob(caseNumber) {
  return caseFilePath(caseNumber, "stylesheet.*");
}

function caseConfig(caseNumber, ext) {
  ext = ext || "json";

  return caseFilePath(caseNumber, `config.${ext}`);
}

function prepResults(results) {
  return results.map(result => {
    // The _postcssResult object is not part of our API and is huge
    const preppedResult = _.omit(result, "_postcssResult");

    // The `source` of each file will not be the same on different machines or platforms
    preppedResult.source = path
      .relative(__dirname, result.source)
      .replace(new RegExp(`\\${path.sep}`, "g"), "/");

    return preppedResult;
  });
}

function stripColors(input) {
  return stripAnsi(input);
}

module.exports = {
  caseFilePath,
  caseStylesheetGlob,
  caseConfig,
  prepResults,
  stripColors
};
