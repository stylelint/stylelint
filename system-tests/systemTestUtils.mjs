import { copyFile, readFile } from 'node:fs/promises';
import { fileURLToPath } from 'node:url';
import os from 'node:os';
import path from 'node:path';
import process from 'node:process';

import replaceBackslashes from '../lib/testUtils/replaceBackslashes.mjs';
import uniqueId from '../lib/testUtils/uniqueId.mjs';

const __dirname = fileURLToPath(new URL('.', import.meta.url));

export function caseFilePath(caseNumber, fileName = 'stylesheet') {
	return replaceBackslashes(path.join(__dirname, caseNumber, fileName));
}

export function caseFiles(caseNumber) {
	return caseFilePath(caseNumber, 'stylesheet.*');
}

export function caseConfigFile(caseNumber, ext = 'json') {
	return caseFilePath(caseNumber, `config.${ext}`);
}

export async function caseConfig(caseNumber) {
	return JSON.parse(await readFile(caseConfigFile(caseNumber, 'json'), 'utf8'));
}

export async function caseCode(caseNumber, ext = 'css') {
	return readFile(`${caseFilePath(caseNumber)}.${ext}`, 'utf8');
}

export async function caseFilesForFix(caseNumber, ext = 'css') {
	const tempPath = replaceBackslashes(path.join(os.tmpdir(), `stylesheet-${uniqueId()}.${ext}`));

	await copyFile(path.join(__dirname, caseNumber, `stylesheet.${ext}`), tempPath);

	return tempPath;
}

export function prepForSnapshot({ results, cwd, output, report, ...rest }) {
	// If output isn't fixed code
	if (output.startsWith('[')) {
		// The `source` of each file varies between platforms or if a tmp file is used
		output = JSON.parse(output).map((warning) => ({
			...warning,
			source: '/path/to/dummy.css',
		}));
	}

	if (report) {
		// The `source` of each file varies between platforms or if a tmp file is used
		report = JSON.parse(report).map((warning) => ({
			...warning,
			source: '/path/to/dummy.css',
		}));
	}

	return {
		cwd: path.relative(process.cwd(), cwd),
		// The _postcssResult object is not part of our API and is huge
		results: results.map((result) => {
			delete result._postcssResult;

			return { ...result, source: '/path/to/dummy.css' };
		}),
		output, // TODO: Deprecated. Remove in the next major version.
		report,
		...rest,
	};
}
