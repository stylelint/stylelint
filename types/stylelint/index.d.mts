export {
	default,
	BuiltInRules,
	Config,
	ConfigRuleSettings,
	CosmiconfigResult,
	CssSyntaxError,
	CustomSyntax,
	DisableOptions,
	DisableOptionsReport,
	DisablePropertyName,
	DisableReportRange,
	DisabledRange,
	DisabledRangeObject,
	DisabledWarning,
	Formatter,
	FormatterType,
	Formatters,
	GetLintSourceOptions,
	GetPostcssOptions,
	InternalApi,
	LintResult,
	LinterOptions,
	LinterResult,
	LonghandSubPropertiesOfShorthandProperties,
	Plugin,
	PostcssPluginOptions,
	PostcssResult,
	Problem,
	PublicApi,
	Rule,
	RuleBase,
	RuleContext,
	RuleMessage,
	RuleMessages,
	RuleMeta,
	RuleOptions,
	RuleOptionsPossible,
	RuleSeverity,
	Severity,
	ShorthandProperties,
	StylelintPostcssResult,
	Utils,
	Warning,
	WarningOptions,
} from './index.js';
