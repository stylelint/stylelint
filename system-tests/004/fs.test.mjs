import stylelint from '../../lib/index.mjs';

import { caseConfigFile, caseFilesForFix, prepForSnapshot } from '../systemTestUtils.mjs';

const CASE_NUMBER = '004';

it('fs - errored state for reportNeedlessDisables', async () => {
	expect(
		prepForSnapshot(
			await stylelint.lint({
				files: await caseFilesForFix(CASE_NUMBER),
				configFile: caseConfigFile(CASE_NUMBER),
				reportNeedlessDisables: true,
			}),
		),
	).toMatchSnapshot();
}, 10000);

it('fs - no errored state', async () => {
	expect(
		prepForSnapshot(
			await stylelint.lint({
				files: await caseFilesForFix(CASE_NUMBER),
				configFile: caseConfigFile(CASE_NUMBER),
			}),
		),
	).toMatchSnapshot();
}, 10000);
