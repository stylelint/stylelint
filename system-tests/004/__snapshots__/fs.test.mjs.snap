// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fs - errored state for reportNeedlessDisables 1`] = `
{
  "cwd": "",
  "errored": true,
  "output": [
    {
      "deprecations": [],
      "errored": false,
      "invalidOptionWarnings": [],
      "parseErrors": [],
      "warnings": [
        {
          "column": 1,
          "endColumn": 38,
          "endLine": 1,
          "line": 1,
          "rule": "--report-needless-disables",
          "severity": "error",
          "text": "Needless disable for "block-no-empty"",
        },
      ],
    },
  ],
  "reportedDisables": [],
  "results": [
    {
      "deprecations": [],
      "errored": false,
      "ignored": undefined,
      "invalidOptionWarnings": [],
      "parseErrors": [],
      "warnings": [
        {
          "column": 1,
          "endColumn": 38,
          "endLine": 1,
          "line": 1,
          "rule": "--report-needless-disables",
          "severity": "error",
          "text": "Needless disable for "block-no-empty"",
        },
      ],
    },
  ],
  "ruleMetadata": {
    "block-no-empty": {
      "url": "https://stylelint.io/user-guide/rules/block-no-empty",
    },
  },
}
`;

exports[`fs - no errored state 1`] = `
{
  "cwd": "",
  "errored": false,
  "output": [
    {
      "deprecations": [],
      "errored": false,
      "invalidOptionWarnings": [],
      "parseErrors": [],
      "warnings": [],
    },
  ],
  "reportedDisables": [],
  "results": [
    {
      "deprecations": [],
      "errored": false,
      "ignored": undefined,
      "invalidOptionWarnings": [],
      "parseErrors": [],
      "warnings": [],
    },
  ],
  "ruleMetadata": {
    "block-no-empty": {
      "url": "https://stylelint.io/user-guide/rules/block-no-empty",
    },
  },
}
`;
